### PR TITLE
refactor(kuma-cp): refactor retry configurer

### PR DESCRIPTION
**Pull request by lukidzi**
I am not sure if that is the best way, if you think we could make it more configurable I am happy to change it.

* changed retry configurer to return configuration which might be used by a different component
* gateway/virtual Hosts/ routes actions are now using the same code
* removed some duplication of the code from gateway
* removed some unnecessary code that was responsible for the creation vh for route and moved it into the builder

- [X] [Link to relevant issue][1] as well as docs and UI issues -- fix: #3140 
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/xds/envoy/routes/v3/retry_configurer.go [hunk 1]
```diff
@@ -150,33 +143,20 @@ func GrpcRetryOn(conf []mesh_proto.Retry_Conf_Grpc_RetryOn) string {
...
-			virtualHost.RetryPolicy = policy
-		}
-
+	var policy *envoy_route.RetryPolicy
+	switch protocol {
+	case "http":
+		policy = genHttpRetryPolicy(retry.Spec.Conf.GetHttp())
+	case "grpc":
+		policy = genGrpcRetryPolicy(retry.Spec.Conf.GetGrpc())
+	default:
```
Comment (bartsmykla) [inline pkg/xds/envoy/routes/v3/retry_configurer.go:156]:
```suggestion

	switch protocol {
	case "http":
		return genHttpRetryPolicy(retry.Spec.Conf.GetHttp())
	case "grpc":
		return genGrpcRetryPolicy(retry.Spec.Conf.GetGrpc())
	default:
```

File: pkg/xds/envoy/routes/v3/retry_configurer.go [hunk 2]
```diff
@@ -150,33 +143,20 @@ func GrpcRetryOn(conf []mesh_proto.Retry_Conf_Grpc_RetryOn) string {
...
 		return nil
 	}
-
-	return UpdateHTTPConnectionManager(filterChain, updateFunc)
+	return policy
```
Comment (bartsmykla) [inline pkg/xds/envoy/routes/v3/retry_configurer.go:159]:
```suggestion
```

File: pkg/xds/envoy/listeners/filter_chain_configurers.go [hunk 3]
```diff
@@ -236,11 +237,12 @@ func Retry(
 	if retry == nil {
 		return FilterChainBuilderOptFunc(nil)
 	}
-
-	return AddFilterChainConfigurer(&v3.RetryConfigurer{
-		Retry:    retry,
-		Protocol: protocol,
-	})
+	return AddFilterChainConfigurer(
```
Comment (bartsmykla) [inline pkg/xds/envoy/listeners/filter_chain_configurers.go:240]:
```suggestion

	return AddFilterChainConfigurer(
```

**Review by bartsmykla — APPROVED** [review]
Just few suggested nits

File: pkg/xds/envoy/routes/v3/retry_configurer.go [hunk 4]
```diff
@@ -150,33 +143,19 @@ func GrpcRetryOn(conf []mesh_proto.Retry_Conf_Grpc_RetryOn) string {
...
-			policy = genGrpcRetryPolicy(c.Retry.Spec.Conf.GetGrpc())
-		default:
-			return nil
-		}
-
-		for _, virtualHost := range manager.GetRouteConfig().VirtualHosts {
-			virtualHost.RetryPolicy = policy
-		}
-
+	var policy *envoy_route.RetryPolicy
```
Comment (bartsmykla) [inline pkg/xds/envoy/routes/v3/retry_configurer.go:150]:
```suggestion

```

File: pkg/xds/envoy/routes/v3/retry_configurer.go [hunk 5]
```diff
@@ -150,13 +150,12 @@ func RetryConfig(retry *core_mesh.RetryResource, protocol core_mesh.Protocol) *e
 	var policy *envoy_route.RetryPolicy
```
Comment (michaelbeaumont) [inline pkg/xds/envoy/routes/v3/retry_configurer.go]:
```suggestion
```

**Pull request merged**